### PR TITLE
[780] Allow admin to edit admin users

### LINKED ISSUE #780: Admin can edit other admin accounts
Part of epic #4

**What type of user is this for? [admin/volunteer/supervisor/all]**
admin

**Where does/should this occur?**
On the admin dashboard, at the bottom underneath subheader "Admins". 

<a href="https://casa-r4g-staging.herokuapp.com/">Log in to staging</a> and use these credentials to access the admin dashboard:
username – casa_admin1@example.com
password – 123456

**Description**
Underneath "Actions" in the Admins section of the Admin dashboard, there should be hyperlinked text per row "Edit" that leads to an Edit Admin user page. This page should look a lot like the <a href="https://casa-r4g-staging.herokuapp.com/users/edit">Edit Profile page for Admins</a>, minus the ability to change password.

Relates to issues: #774, #779

### PR DESCRIPTION
**Pull request by drborges**
### What github issue is this PR for, if any?

Resolves #780 

### What changed, and why?

1. Admins can now edit the `email` and `display_name` of other casa admin users;
2. A new resource was introduced `/casa_admins`, for routes `edit` and `update`;
3. A few `request`, `view`, and `feature` specs were introduced to cover the new code execution path.

### How will this affect user permissions?

- Volunteer permissions: Can't access the edit form.
- Supervisor permissions: Can't access the edit form.
- Admin permissions: Only admins will have access to the edit form.

### How is this tested? (please write tests!) 💖💪

For manual tests, one should:

Happy path:

1. Log in as an admin user;
2. On the dashboard page, scroll down to the `Admins` section;
3. Click on the **new** `Edit` link next to a casa admin user listed in that section;
4. Update the user with new information and click `Update`;
5. Verify that you are redirected back to the dashboard page, with a success message on the top of the page.

Specs were added to cover a few scenarios including validations:

**Scenario 1**

```
Given that I am logged into CASA as an admin user
And I am on the dashboard page
I should be able to click on the `Edit` link next to another admin user listed under the `Admins` section
And be taken to an edit admin user form
```

**Scenario 2**

```
Given that I am logged into CASA as an admin user
And I am on the edit admin form page
I should be able to edit the admin's email and display name but not their password
```

**Scenario 3**

```
Given that I am logged into CASA as an admin user
And I am on the edit admin form page
When attempting to update the user without an email or display name
Then I should see validation error messages rendered on the form along with the visual indicators on the form fields.
```

### Feelings gif (optional)

success

### CODE REVIEW

File: app/views/users/edit.html.erb [hunk 1]
```diff
@@ -6,42 +6,42 @@
...
+      <%= form.submit "Update Profile", class: "btn btn-primary" %>
+      <%= link_to "Back", root_path, class: "btn btn-primary" %>
+      <button class="btn btn-warning accordion" id="accordionExample" type="button" data-toggle="collapse" data-target="#collapseOne" aria-expanded="true" aria-controls="collapseOne">
+        Change Password
+      </button>
+    </div>
+    <% end %>
+    <br>
+    <div id="collapseOne" class="collapse" aria-labelledby="headingOne" data-parent="#accordionExample">
+      <%= form_for(@user, as: :user, :url => { :action => "update_password" } ) do |f| %>
```
Comment (jmkoni) [inline app/views/users/edit.html.erb:31]: minor and non-blocking, but should be `url: { action: `

**Review by jmkoni — APPROVED** [review]
Just one non-blocking change requested. Looks good overall :)

File: app/views/users/edit.html.erb (hunk 1 above)
Comment (drborges) [reply app/views/users/edit.html.erb:31]: 🤦🏻‍♂️ copy&paste strikes again 😂 I'll make the change, thank you.

File: spec/requests/casa_admins_spec.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        get edit_casa_admin_path(create(:casa_admin))
+
+        expect(response).to redirect_to root_path
+        expect(response.request.flash[:notice]).to eq "Sorry, you are not authorized to perform this action."
+      end
+    end
+
+    context "unauthenticated request" do
+      it "can successfully access a casa admin edit page" do
```
Comment (gmfvpereira) [inline spec/requests/casa_admins_spec.rb:27]: This message sounds wrong, an unauthenticated request should not access the casa admin page.

File: app/views/users/edit.html.erb (hunk 1 above)
Comment (drborges) [reply app/views/users/edit.html.erb:31]: oh, actually that was just formatting changes. I'll make the update.

File: spec/requests/casa_admins_spec.rb (hunk 2 above)
Comment (drborges) [reply spec/requests/casa_admins_spec.rb:27]: good catch! now that's the copy&pasta issue 😆

File: app/controllers/casa_admins_controller.rb [hunk 3]
```diff
@@ -35,14 +35,13 @@ def set_admin
...
-
-  def casa_admin_values
-    { password: SecureRandom.hex(10), casa_org_id: current_user.casa_org_id }
+    CasaAdminParameters.new(params).with_only(:email, :display_name)
   end
 
-  def casa_admin_params
-    params.require(:casa_admin).permit(:display_name, :email, :password, :casa_org_id)
+  def create_casa_admin_params
+    CasaAdminParameters.new(params)
```
Comment (drborges) [inline app/controllers/casa_admins_controller.rb:42]: @jmkoni can I get your eyes on this change as well? I noticed this `*Parameters` pattern being used in other places and figured that would fi this scenario nicely.

Is this how we'd like to handle strong params moving forward?

cc @derricklannaman, @stevelandwehr.

File: app/views/casa_admins/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    <div class="field form-group">
+      <%= form.label :display_name %>
+      <%= form.text_field :display_name, class: "form-control" %>
+    </div>
+
+    <div class="actions">
+      <%= form.submit "Update", class: "btn btn-primary" %>
+      <%= link_to "Back", root_path, class: "btn btn-primary" %>
+    </div>
```
Comment (ds29) [inline app/views/casa_admins/edit.html.erb:22]: Should we wrap this up in a partial for `_form.html.erb` and share the logic with `new.html.erb` and `edit.html.erb`?

**Review by ds29 — APPROVED** [review]

File: app/views/casa_admins/edit.html.erb (hunk 4 above)
Comment (drborges) [reply app/views/casa_admins/edit.html.erb:22]: yes, I'm working on that now 👍🏼

File: app/controllers/casa_admins_controller.rb (hunk 3 above)
Comment (derricklannaman) [reply app/controllers/casa_admins_controller.rb:42]: @drborges Is this way still whitelisting the acceptable incoming params from the form?

Comment (drborges) [reply app/controllers/casa_admins_controller.rb:42]: @derricklannaman yes. I also ran a quick smoke test to make sure the create functionality still works in addition to the system spec passing 👍

**Comment by Zrrrpy** [comment]
Tested and works on the frontend! Will deploy from QA to staging later.

**Pull request merged**